src/main/java/hudson/remoting/Engine.java

```diff
@@ -496,6 +517,116 @@ public void run() {
         }
     }
 
+    @SuppressFBWarnings(value = "REC_CATCH_EXCEPTION", justification = "checked exceptions were a mistake to begin with")
+    private void runWebSocket() {
+        try {
+            AtomicReference<Channel> ch = new AtomicReference<>();
+            String localCap = new Capability().toASCII();
+            class HeaderHandler extends ClientEndpointConfig.Configurator {
+                Capability remoteCapability = new Capability();
+                @Override
+                public void beforeRequest(Map<String, List<String>> headers) {
+                    headers.put(JnlpConnectionState.CLIENT_NAME_KEY, Collections.singletonList(slaveName));
+                    headers.put(JnlpConnectionState.SECRET_KEY, Collections.singletonList(secretKey));
+                    headers.put(Capability.KEY, Collections.singletonList(localCap));
+                    // TODO use JnlpConnectionState.COOKIE_KEY somehow (see EngineJnlpConnectionStateListener.afterChannel)
+                    LOGGER.fine(() -> "Sending: " + headers);
+                }
+                @Override
+                public void afterResponse(HandshakeResponse hr) {
+                    LOGGER.fine(() -> "Receiving: " + hr.getHeaders());
+                    List<String> remotingMinimumVersion = hr.getHeaders().get("X-Remoting-Minimum-Version");
+                    if (remotingMinimumVersion != null && !remotingMinimumVersion.isEmpty()) {
+                        VersionNumber minimumSupportedVersion = new VersionNumber(remotingMinimumVersion.get(0));
+                        VersionNumber currentVersion = new VersionNumber(Launcher.VERSION);
+                        if (currentVersion.isOlderThan(minimumSupportedVersion)) {
+                            // TODO these errors should trigger a connection close
```

Comment:
This one is tricky since `afterResponse` is not allowed to throw an exception or otherwise cancel the connection. Maybe `onOpen` could track the condition and immediately close the new connection, or maybe a `RuntimeException` works here. Can check into it.